File: .cursor/hooks/before_shell_execution.py
#!/usr/bin/env python3
"""
Cursor hook: beforeShellExecution
Fires before Cursor's agent (or any subagent with terminal access) runs a
shell command. This is the actual enforcement behind ADR-014's "no agent
merges, ever" rule — not a line in a subagent's prompt, a command-level
block that no subagent can reason or be prompted past.

Payload/response shape follows the documented pattern at
https://cursor.com/docs/agent/hooks — re-check field names there if
Cursor's schema has moved since this was written; hooks read a JSON object
from stdin and print a JSON object to stdout.
"""
import json
import re
import sys

BLOCKED_PATTERNS = [
    r"\bgit\s+merge\b",
    r"\bgit\s+push\b[^\n]*\b(main|master)\b",
    r"\bgh\s+pr\s+merge\b",
]


def main() -> None:
    try:
        payload = json.load(sys.stdin)
    except json.JSONDecodeError:
        payload = {}

    command = payload.get("command", "")

    for pattern in BLOCKED_PATTERNS:
        if re.search(pattern, command):
            print(json.dumps({
                "permission": "deny",
                "agentMessage": (
                    f"Blocked: '{command}' matches a merge/direct-push pattern. "
                    "Per ADR-014, no agent in this pipeline may merge or push "
                    "directly to main/master. Use pr-drafter to open a draft PR "
                    "and let a human merge it."
                ),
            }))
            return

    print(json.dumps({"permission": "allow"}))


if __name__ == "__main__":
    main()

File: src/observability/before-shell-execution.ts
/**
 * Cursor hook: beforeShellExecution
 * Blocks merge/direct-push commands. Emits a blocked-command event to Langfuse.
 */
import { getLangfuse, getTrace, flush, readStdin } from './langfuse-client.js';

const BLOCKED = [
  /\bgit\s+merge\b/,
  /\bgit\s+rebase\b/,
  /\bgit\s+push\b[^\n]*\b(--force|--force-with-lease|-f)\b/,
  /\bgit\s+push\b[^\n]*\b(main|master|release)\b/,
  /\bgh\s+pr\s+merge\b/,
  /\bgh\s+api\b[^\n]*\b\/merges?\b/,
];

const raw     = await readStdin();
const payload = JSON.parse(raw || '{}') as { command?: string; conversation_id?: string };
const command = payload.command ?? '';
const convId  = payload.conversation_id ?? 'unknown';
const match   = BLOCKED.find(p => p.test(command));

if (match) {
  const lf    = getLangfuse();
  const trace = getTrace(lf, convId);
  trace?.event({ name: 'blocked-command', input: { command }, level: 'WARNING' });
  await flush(lf);

  process.stdout.write(JSON.stringify({
    permission: 'deny',
    agentMessage: `Blocked: '${command}' matches a merge/direct-push pattern. Per ADR-014, no agent may merge or force-push; this regex list is defense-in-depth behind pr-drafter's missing terminal, not a complete sandbox. Use pr-drafter and let a human merge.`,
  }));
} else {
  process.stdout.write(JSON.stringify({ permission: 'allow' }));
}
